### ISSUE TITLE
Consider ResourceQuota when admitting workloads

### ISSUE THREAD

**Issue by tenzen-y** [kubernetes-sigs/kueue#696]
**What would you like to be added**:
Batch administrators can install [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/) to set constraints for total requests in the namespace.

We should consider ResourceQuotas when admitting Workloads.

**Why is this needed**:
~~For now, we may sometime face the following problems:~~

~~1. if we use `Sequential Admission with Ready Pods` and ResourceQuotas together, we may face deadlocks.~~
~~2. Many unschedulable pods (with pending status) could be created.~~

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [ ] Docs update

The artifacts should be linked in subsequent comments.

**Comment by alculquicondor** [comment]
I like this idea and I don't think we will face the problems you mention.
When ResourceQuotas are hit, the effect is that the pod can't even be created. The scheduler doesn't even get to see those pods. And because we will only allow jobs that satisfy the resource quota to be created, they won't hit the limit.
However, the open question is what to do with flavors. For it's simplest form, probably there would only be one flavor for which all resources are defined. And I guess we would create a ClusterQueue for each ResourceFlavor. It could be an optional mode of operation.

**Comment by tenzen-y** [comment]
> However, the open question is what to do with flavors. For it's simplest form, probably there would only be one flavor for which all resources are defined. And I guess we would create a ClusterQueue for each ResourceFlavor.

Uhm. In that case, we can not create ClusterQueues with multiple ResourceFlavors if batch admins select the optional mode of operation, right?

**Comment by alculquicondor** [comment]
Unless there are some annotations in the ResoureQuota, but that could be a later extension

**Comment by tenzen-y** [comment]
> Unless there are some annotations in the ResoureQuota, but that could be a later extension

Sounds good.

**Comment by samzong** [comment]
> Unless there are some annotations in the ResoureQuota, but that could be a later extension

Could you explain this design in more detail? Sorry to bother you after such a long time.

How to balance clusterqueue quota vs. namespace ResourceQuota is a bit complicated, especially when clusterqueue is a cluster-level resource.

**Comment by kerthcet** [comment]
Would NS ResourceQuota be another admission check?

**Comment by panpan0000** [comment]
one of the import use case as below:

In a namespace, there may be different types of workloads(I call it "hybrid-workload") consuming hardware accelerators:

- Batch Jobs
- interactive notebook
- model serving deployments




- `kueue` can management the batch jobs, to limit them under quota of the ClusterQueue's FlavorQuotas.
- But the other workloads are still controlled by native kubernetes namespaced [resourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/)

the challenges are :

- The two quota-systems are in different dimension, FlavorQuota are not under namespace granularity, it's not suitable for multi-tenant scenarios (maybe I'm wrong..)
- The two kinds of workloads actually compete for same hardware resources. But they are not aware of the other. For example, maybe jobs eat up all physical GPU, and later a notebook fails to launch but plenty of namespaced-resource-quota are still available.


In short, I think it's valuable to address the challenges , so that user can leverage `kueue` in multi-tenant circumstance with  
"hybrid-workload" :-)

**Comment by tenzen-y** [comment]
> Would NS ResourceQuota be another admission check?

@kerthcet I think so too. Additional admission checks would resolve this issue.
@alculquicondor Any concerns?

**Comment by tenzen-y** [comment]
I'm not against supporting ResourceQuota by admission checks, and I agree with the ResourceQuota support. I'm sure there are use cases in which we want to manage kueue + ResourceQuota, although, in an ideal world, all workloads are managed by the kueue's flavorQuotas.

**Comment by tenzen-y** [comment]
Also, we need to support elastic jobs to support serving deployment with autoscaling semantics.

https://github.com/kubernetes-sigs/kueue/issues/77

**Comment by tenzen-y** [comment]
> So admission check might be the simplest thing we can boot with.

I agree. An additional admission check would be better.
As another use case, existing resources managed by ResourceQuotas could easily adapt to the kueue.

In the adaption to the existing environments, I think that the ResourceQuota support has an advantage in terms of using all kueue features over using naked pod integration if our naked pod integration doesn't support all kueue features.

However, I'm on the fence about whether we should support ResourceQuota by AdmissionCheck since using ResourceQuota is a temporary measure since as I mentioned above, ideally, all resources are managed by Kueue's flavorQuotas.

@alculquicondor What do you think about supporting ResourceQuota by AdmissionCheck for easy adaptation to the existing environment?

**Comment by alculquicondor** [comment]
I don't think we should implement it.

As explained, it will be possible for someone to implement it using AdmissionChecks, but I don't think we should support it  in core Kueue. I would accept a controller in the cmd/experimental package though.

**Comment by tenzen-y** [comment]
> I don't think we should implement it.
> 
> As explained, it will be possible for someone to implement it using AdmissionChecks, but I don't think we should support it in core Kueue. I would accept a controller in the cmd/experimental package though.

The `cmd/experimental` sounds much more reasonable. SGTM

**Comment by kerthcet** [comment]
/remove-lifecycle rotten
Still valid I think

**Comment by yuvipanda** [comment]
This would still be very helpful for multitenant scenarios, where there are multiple users of interactive computing on the same cluster and need access to individual queues.
